posthog/api/dashboard.py

```diff
@@ -216,15 +219,34 @@ def update(self, instance: Dashboard, validated_data: Dict, *args: Any, **kwargs
 
         validated_data.pop("use_template", None)  # Remove attribute if present
 
+        being_undeleted = instance.deleted and "deleted" in validated_data and not validated_data["deleted"]
+        if being_undeleted:
+            DashboardTile.objects.filter(dashboard__id=instance.id).update(deleted=False)
+            insights_to_undelete = []
+            for tile in DashboardTile.objects.filter(dashboard__id=instance.id):
+                if tile.insight and tile.insight.deleted:
+                    tile.insight.deleted = False
+                    insights_to_undelete.append(tile.insight)
+            Insight.objects.bulk_update(insights_to_undelete, ["deleted"])
+
         initial_data = dict(self.initial_data)
 
         instance = super().update(instance, validated_data)
 
         if validated_data.get("deleted", False):
-            DashboardTile.objects.filter(dashboard__id=instance.id).delete()
+            if self.validated_data.get("delete_insights", False):
+                insights_to_update = []
+                for insight in Insight.objects.filter(dashboard_tiles__dashboard=instance.id):
+                    if insight.dashboard_tiles.exclude(deleted=True).count() == 1:
+                        insight.deleted = True
+                        insights_to_update.append(insight)
+
+                Insight.objects.bulk_update(insights_to_update, ["deleted"])
```

Comment:
This is a bit hard to read inside `update()`. Could it be something like `instance.delete_related_insights()` and `instance.undo_delete_related_insights()`? The `Dashboard` class seems like a pretty good place for this logic, but up to you, serializer method works too

Reply:
Hmmm, nicer in the Dashboard class but circular import warnings so I'll leave things in the serializer...